Network/Socket/Posix/Cmsg.hsc

```diff
@@ -87,24 +88,27 @@ filterCmsg cid cmsgs = filter (\cmsg -> cmsgId cmsg == cid) cmsgs
 --   Each control message type has a numeric 'CmsgId' and a 'Storable'
 --   data representation.
 class Storable a => ControlMessage a where
-    controlMessageId :: a -> CmsgId
+    controlMessageId :: Proxy a -> CmsgId
```

Comment:
You don't even need the Proxy argument if we drop the support for GHC 8.0 (cf. https://ryanglscott.github.io/2019/02/06/proxy-arguments-in-class-methods/). In general, polymorphic proxy type in class methods is not necessarily a good idea since it prevents GeneralisedNewtypeDeriving but I guess it doesn't matter here.

Reply:
Did you mean "drop the support for GHC 7.x"?

Reply:
I did mean GHC 8.0. GHC 8.0 doesn't support the following definition even with `AllowAmbiguousTypes`:

```haskell
class C a where
   foo :: Int
```

Reply:
I read the blog post, and it is quite interesting material, but I don't see a compelling need for newtype deriving with socket Control message structures, so we can go with the abstract `proxy` without fear, but the concrete `Proxy` is perhaps in fact less restrictive.  The deriving issues don't come up, and a programmer passing `undefined :: Proxy a` rather than `Proxy :: Proxy a` deserves any resulting breakage.

So I'd keep it simple and stick with `Proxy a` and not yet require the fancier techniques that need much more recent advanced GHC features.

Reply:
@fumieval  Ryan's blog says that this is supported from GHC 8.0. And the code with this approach can be complied with GHC 8.0.2 on my local machine.

Reply:
Which `this` are you referring to?

Reply:
@vdukhovni  I would like to get rid of `undefined` as much as possible.

Reply:
Yes, that's clear, no `undefined.  But the question is whether you're talking about `Proxy a -> Int`, `proxy a -> Int` or just `Int` (requiring `AmbiguousTypes` and `TypeApplications`, and some recent GHC).

Reply:
Hmm right, I was misremembering

Reply:
It does look like GHC 8.0 supports the simple signature with no argument.  Running `stack ghci --resolver lts-7.24` I get GHC 8.0.1, which handles the form with no arguments:
```
GHCi, version 8.0.1: http://www.haskell.org/ghc/  :? for help
Prelude> :set -XTypeApplications
Prelude> :set -XAllowAmbiguousTypes
Prelude> class C a where { foo :: Int }
Prelude> instance C Int where { foo = 0 }
Prelude> instance C Char where { foo = 1 }
Prelude> foo @Int
0
Prelude> foo @Char
1
```

Reply:
But, it breaks `GeneralizedNewtypeDeriving` with GHC prior to 8.8.  Do we care?
Here's the error with 8.6.5:
```
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Prelude> :set -XTypeApplications
Prelude> :set -XAllowAmbiguousTypes
Prelude> :set -XGeneralizedNewtypeDeriving 
Prelude> class C a where { foo :: Int }
Prelude> instance C Int where { foo = 0 }
Prelude> instance C Char where { foo = 1 }
Prelude> foo @Int
0
Prelude> foo @Char
1
Prelude> newtype Bar = Bar Int deriving(C)

<interactive>:9:32: error:
    • Ambiguous type variable ‘a0’ arising from a use of ‘foo’
      prevents the constraint ‘(C a0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘a0’ should be.
      These potential instances exist:
        instance C Bar -- Defined at <interactive>:9:32
        instance [safe] C Char -- Defined at <interactive>:6:10
        instance [safe] C Int -- Defined at <interactive>:5:10
    • In the third argument of ‘GHC.Prim.coerce’, namely ‘foo’
      In the expression: GHC.Prim.coerce @Int @Int foo :: Int
      In an equation for ‘foo’:
          foo = GHC.Prim.coerce @Int @Int foo :: Int
      When typechecking the code for ‘foo’
        in a derived instance for ‘C Bar’:
        To see the code I am typechecking, use -ddump-deriv
```
and success with 8.8.3:
```
GHCi, version 8.8.3: https://www.haskell.org/ghc/  :? for help
Prelude> :set -XGeneralizedNewtypeDeriving 
Prelude> :set -XTypeApplications
Prelude> :set -XAllowAmbiguousTypes
Prelude> class C a where { foo :: Int }
Prelude> instance C Int where { foo = 0 }
Prelude> instance C Char where { foo = 1 }
Prelude> foo @Int
0
Prelude> foo @Char
1
Prelude> newtype Bar = Bar Int deriving(C)
Prelude> foo @Bar
0
```

Reply:
OK. GHC 8.0 is supported. So, I prefer to `controlMessageId :: CmsgId` with `AllowAmbiguousTypes` and `TypeApplications`.

Reply:
Do we really need `deriving(ControlMessage)`? I don't think so.

Reply:
Works for me.  I don't think that lack of support for newtype deriving in GHC 8.0&ndash;8.6 is a significant drawback, especially since `Cmsg` is a new interface with no legacy users.